Changelog: append a generated list of first time contributors by PR label

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

This PR attempts to address the first few suggestions on [this discussion](https://github.com/WordPress/gutenberg/discussions/38244) regarding Establish a first-time contributor props culture in Gutenberg.

It's pretty basic and requires user intervention in terms of applying the correct labels in Github. 

The PR aims to:

- find all PRs merged with the label `First-time contributor`.
- find the user who created the PR.
- add them to a list of contributors under the title `First time contributors`.

This list is appended to the bottom of the existing Changelog and looks like this:

```
## First time contributors

The following PRs were merged by first time contrbutors:

- @alanjacobmathew: [Type] Documentation  Updating Faq.md typo. ([38154](https://github.com/WordPress/gutenberg/pull/38154))
- @imangm: RangeControl: Fix space between icons and rail. ([36935](https://github.com/WordPress/gutenberg/pull/36935))
- @jorgecontreras: Fix draft previews. ([37952](https://github.com/WordPress/gutenberg/pull/37952))
- @knilkantha: Scripts: Corrected "npm start:Hot" in the README file. ([38178](https://github.com/WordPress/gutenberg/pull/38178))
```

I envisage the release lead confidence checking all the PRs in the list (note it's generally fairly small) and then amending the list as required. I also envisage us creating a culture of ensuring PRs are labelled accordingly in order to give kudos to new contributors.

I acknowledge that this doesn't cover other types of contribution such as reviews...etc. However the concept could be extended as suggested in the discussion, namely by creating labels for `First-time reporter` and `First-time commenter` to track non-PR contributions as well (although that would require further tweaks to this script). I believe however that this PR is a good first step towards establishing a regular culture of recognising first time contributors. In short, it's not perfect, but it's **better than what we have right now**.

## Testing Instructions

1. Checkout PR.
2. Open terminal in Gutenberg repo.
3. Run `npm run changelog -- --milestone="Gutenberg 12.5"` (or use milestone of your choosing!).
4. See changelog generated with FTC section appended to the bottom.
5. Check the FTC's match up with [the labels in Github](https://github.com/WordPress/gutenberg/pulls?q=is%3Apr+milestone%3A%22Gutenberg+12.5%22+label%3A%22First-time+Contributor%22) - note I've filtered out "bots" from the list as they don't require any validation or recognition (yet 😉  ).

## Screenshots <!-- if applicable -->

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave committed:
Create list of first time contributors by PR label

gziolo commented on bin/plugin/commands/changelog.js:
```diff
@@ -757,6 +757,76 @@ function sortFeatureGroups( featureGroups ) {
+		return pr.labels.find(
+			( { name } ) => name.toLowerCase() === 'first-time contributor'
+		);
+	} );
```
When I talked about this idea with @priethor, he mentioned that sometimes users open more than one PR and since they don't have any commits all of them get this label. That's why it would be good to ensure that the list of users is unique for a given release. Maybe we need to improve the GitHub action that applies the label, to ensure it correctly removes it from the PR at the time of merging if that is a subsequent PR.

getdave replied:
Sounds good. Where do these actions live? I found the `workflows/` dir in the repo but I couldn't see anything about `First-time contributor`.

getdave replied:
How about I simply group the PRs by username? That way if the user has more than 1 PR merged in this milestone then they get the credits for that only.

priethor replied:
> he mentioned that sometimes users open more than one PR and since they don't have any commits all of them get this label
 
Indeed, there are some cases where a contributor would open a second PR without the first one being closed, so the second one gets the `First-time contributor` label assigned, too. As far as I can tell, only one `First-time contributor`-labeled PR per contributor has been merged; I wouldn't be too worried about this edge case. It is very likely that when a new contributor opens multiple PRs, they are trying to solve the same problem in different ways, rather than trying to fix multiple things as soon as they start contributing. Probably filtering unique contributors in the same release will work because of this reason.

gziolo replied:
There is an npm package with some custom automations:

https://github.com/WordPress/gutenberg/tree/trunk/packages/project-management-automation

It’s definitely nice to have and grouping by username should be perfect for start like you both pointed out.

I was mostly contemplating on the current state of the automation that applies all those labels 😃

getdave replied:
> There is an npm package with some custom automations:
>
> https://github.com/WordPress/gutenberg/tree/trunk/packages/project-management-automation

That's super cool. Diving into that now 👀 

> It’s definitely nice to have and grouping by username should be perfect for start like you both pointed out.

If we're happy with the general direction I'll proceed with refactoring and tidying up this PR.

getdave commented on bin/plugin/commands/changelog.js:
```diff
@@ -757,6 +757,76 @@ function sortFeatureGroups( featureGroups ) {
+ *
+ * @return {string} The formatted changelog string.
+ */
+function formatContributors( pullRequests ) {
```
If folks are happy with the general direction I'm happy to refactor this a bit and DRY up the methods.

getdave committed:
Refactor into single purpose methods

getdave committed:
Remove unecessary verbosity

getdave committed:
Extract common Issue/PR title formatting utility

getdave committed:
Normalize title in contributors list

getdave committed:
Add top level changelog title

getdave committed:
Remove duplicate full stop

getdave committed:
Qualify and correct method comment

getdave committed:
Update the snapsot to include the new top level headings

getdave committed:
Add test for getFormattedItemDescription

getdave committed:
Rename functions for consistency with existing

getdave committed:
Add snapshot test for formatContributors

gziolo commented on bin/plugin/commands/changelog.js:
```diff
@@ -757,6 +775,94 @@ function sortFeatureGroups( featureGroups ) {
+		auth: settings.token,
+	} );
+
+	const pullRequests = await fetchAllPullRequests( octokit, settings );
```
It looks like we call the same request twice:
- for changelo
- for new contributors 

How about it gets lifted outside and passed to both functions?

getdave replied:
Yeh I was wondering about that. Ok I'll extract it 👍

getdave replied:
@gziolo I've completed the refactor and in the process possibly uncovered a bug. I was previously resolving two promises and using the `PromiseResult` as the output to be `log`'d.

Now by extracting I've removed the need to have so many async functions and also made the concat'ing of the two formatted strings a lot clearer.

I would appreciate another quick 👀 from you on this as it's changed a fair amount since your original 👍 .

gziolo commented on bin/plugin/commands/test/__snapshots__/changelog.js.snap:
```diff
@@ -207,3 +209,15 @@ exports[`formatChangelog verify that the changelog is properly formatted 1`] = `
+
+The following PRs were merged by first time contrbutors:
+
+- @kapilpaul: Add spaces in add_theme_support documentation code. ([33796](https://github.com/WordPress/gutenberg/pull/33796))
```
Nit: this list could also be sorted by the username 😃

getdave replied:
Yes! Everyone loves sorting 👍

getdave replied:
✅

gziolo reviewed (approved):
Excellent work. It’s good enough to be merged. I left some minor comments with enhancements  to consider.

talldan commented on bin/plugin/commands/changelog.js:
```diff
@@ -757,6 +775,94 @@ function sortFeatureGroups( featureGroups ) {
+ */
+function getFirstTimeContributorPRs( pullRequests ) {
+	return pullRequests.filter( ( pr ) => {
+		if ( pr.user.login === 'dependabot[bot]' ) {
```
It should be possible to make this work for all bots using the `type` field that github provides on the `user` object (`pr.user.type === 'Bot'`).

See https://github.com/WordPress/gutenberg/pull/38393 for reference.

getdave replied:
Ooooo nice. Thanks Dan 🙇‍♂️

getdave replied:
✅

getdave committed:
Sort users alphabetically

Addresses https://github.com/WordPress/gutenberg/pull/38372#discussion_r796289457

getdave committed:
Filter out bots using user type field

Addresses https://github.com/WordPress/gutenberg/pull/38372#discussion_r796333698

getdave committed:
Refactor to single API request and simplify

getdave committed:
Improve function comments

gziolo commented on bin/plugin/commands/changelog.js:
```diff
@@ -757,6 +768,84 @@ function sortFeatureGroups( featureGroups ) {
+ */
+function getContributors( pullRequests ) {
+	const contributorsList = flow( [
+		getFirstTimeContributorPRs,
```
I think you could add a `uniqBy` call from lodash just before sorting to remove duplicates for the same user. It might be as simple as:

```js
( items ) => uniqBy( items, ( item ) => item.user.login ),
```

getdave replied:
✅  and also added a test for the new function

gziolo reviewed (approved):
Everything looks good.

getdave committed:
De-dupe conttributor list by username

gziolo commented:
🚢 🎉

getdave commented:
@fabiankaegy As release lead for the next Plugin release, I thought I'd give you a heads up that this PR will change the format of the changelog. You'll want to be aware when generating. Any problems please do let me know.

getdave commented:
I spoke with @fabiankaegy and it turns out the contributors list didn't get auto-appended to the 12.5 release. Fabian has now generated manually via command line but I'm thinking perhaps Github requires a specific format for the changelog and thus has rejected the FTC section.

talldan commented:
> I spoke with @fabiankaegy and it turns out the contributors list didn't get auto-appended to the 12.5 release. Fabian has now generated manually via command line but I'm thinking perhaps Github requires a specific format for the changelog and thus has rejected the FTC section.

Isn't the changelog only generated at RC stage?

getdave commented:
I thought we generated it again for the release but I could be wrong. I was going to try forking this repo and running a few releases on my fork to see what happens.

talldan commented:
@getdave Looks like it worked for 12.6.0 🎉 

https://github.com/WordPress/gutenberg/releases/tag/v12.6.0-rc.1
